simpeg/utils/mesh_utils.py

```diff
@@ -97,15 +100,192 @@ def surface2inds(vrtx, trgl, mesh, boundaries=True, internal=True):
+    if grid.squeeze().ndim == 1:
+        grid_inds = np.asarray(
+            [np.abs(pt - grid).argmin() for pt in pts.tolist()], dtype=int
+        )
+    else:
+        tree = cKDTree(grid)
+        _, grid_inds = tree.query(pts)
+
+    return grid_inds
+
+
+def get_discrete_topography(mesh, active_cells, option="top"):
+    """
+    Generate discrete topography locations.
+
+    Parameters
+    ----------
+    mesh : TensorMesh or discretize.TreeMesh
+        A tensor or tree mesh.
+    active_cells : numpy.ndarray of bool or int
+        Active cells index; i.e. indices of cells below surface
+    option : {"top", "center"}
+        Use string to specify if the surface passes through the
+        tops or cell centers of surface cells.
+
+    Returns
+    -------
+    (n,) or (n, 1) numpy.ndarray
+        Horizontal locations x[y] for discrete topography.
+    (n,) numpy.ndarray
+        Elevations for discrete topography.
+    """
+    if mesh._meshType == "TENSOR":
+        if mesh.dim == 3:
+            mesh2D = TensorMesh([mesh.h[0], mesh.h[1]], mesh.x0[:2])
+            zc = mesh.cell_centers[:, 2]
+            ACTIND = active_cells.reshape(
+                (mesh.vnC[0] * mesh.vnC[1], mesh.vnC[2]), order="F"
+            )
+            ZC = zc.reshape((mesh.vnC[0] * mesh.vnC[1], mesh.vnC[2]), order="F")
+            topoCC = np.zeros(ZC.shape[0])
+
+            for i in range(ZC.shape[0]):
+                if option == "top":
+                    ind = np.argmax(ZC[i, :][ACTIND[i, :]])
+                    dz = mesh.h[2][ACTIND[i, :]][ind] * 0.5
+                elif option == "center":
+                    dz = 0.0
+                else:
+                    raise ValueError("'option' must be 'top' or 'center'.")
+                topoCC[i] = ZC[i, :][ACTIND[i, :]].max() + dz
+            return mesh2D.cell_centers, topoCC
+
+        elif mesh.dim == 2:
+            mesh1D = TensorMesh([mesh.h[0]], [mesh.x0[0]])
+            yc = mesh.cell_centers[:, 1]
+            ACTIND = active_cells.reshape((mesh.vnC[0], mesh.vnC[1]), order="F")
+            YC = yc.reshape((mesh.vnC[0], mesh.vnC[1]), order="F")
+            topoCC = np.zeros(YC.shape[0])
+            for i in range(YC.shape[0]):
+                ind = np.argmax(YC[i, :][ACTIND[i, :]])
+                if option == "top":
+                    dy = mesh.h[1][ACTIND[i, :]][ind] * 0.5
+                elif option == "center":
+                    dy = 0.0
+                else:
+                    raise ValueError("'option' must be 'top' or 'center'.")
+                topoCC[i] = YC[i, :][ACTIND[i, :]].max() + dy
+            return mesh1D.cell_centers, topoCC
+
+    elif mesh._meshType == "TREE":
+        inds = mesh.get_boundary_cells(active_cells, direction="zu")[0]
+
+        if option == "top":
+            dz = mesh.h_gridded[inds, -1] * 0.5
+        elif option == "center":
+            dz = 0.0
+        return mesh.cell_centers[inds, :-1], mesh.cell_centers[inds, -1] + dz
```

Comment:
building a new mesh for the tree mesh would actually be pretty simple, from this array of indices. Just form a new mesh in D-1, grab the level of each cell from the old mesh, then insert each cell into the new mesh at the same level.